# Контрольная работа
# Вариант 19



# Выполнил:
# Радайкин Артём, ПМ20-4 

## Загружаем данные

In [ ]:
install.packages('lmtest')
library('lmtest')

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

also installing the dependency ‘zoo’


Loading required package: zoo


Attaching package: ‘zoo’


The following objects are masked from ‘package:base’:

    as.Date, as.Date.numeric




In [ ]:
df = read.table('19.txt', header=T, sep='\t', dec='.')
df

## Построение модели


In [ ]:
model_1 = lm(y ~ x1 + x2, data=df)

In [ ]:
summary(model_1)


Call:
lm(formula = y ~ x1 + x2, data = df)

Residuals:
     Min       1Q   Median       3Q      Max 
-1.28332 -0.31502 -0.08199  0.38497  1.72936 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) 16.55736   29.18175   0.567    0.572    
x1           0.73504    0.08737   8.412  2.6e-12 ***
x2           0.07590    0.09904   0.766    0.446    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.6118 on 72 degrees of freedom
Multiple R-squared:  0.6563,	Adjusted R-squared:  0.6467 
F-statistic: 68.74 on 2 and 72 DF,  p-value: < 2.2e-16


Y_hat = 16.55736 + 0.73504 * x1 + 0.0759*x2
       

> (29.18175)    (0.08737)  (0.09904)     

> F-statistic: 68.74

> R-squared (R^2):  0.6563

> Adjusted R-squared (R^2_adj):  0.6467

> Residual standard error (Se): 0.6118

> A_mean = 0.163 % (посчитанно ниже)


**Оцените качество моделей, используя коэффициент детерминации и среднюю 
относительную ошибку аппроксимации.**

In [ ]:
R_2 =  0.656
R_adj = 0.647
y_p = df[,1]
yi_hat = predict(model_1)

A_mean = (1/75) * (sum(abs((y_p - yi_hat) / y_p))) * 100
paste0('A_mean = ', round(A_mean,3), ' %' )

[1] "A_mean = 0.163 %"

**Проведём тесты на значимость параметров моделей и моделей в целом.**

In [ ]:
t_alfa = 0.567
t_b1 = 8.412
t_b2 = 0.766
t_fr = qt(0.975, 72) # t_табл(0.05, 72) 
paste0(round(t_fr,3))

[1] "1.993"

>  |t_b1| > t_табл => гипотеза H0 отвергается => параметры модели значимы

> |t_alfa|, |t_b2| < t_табл => гипотеза H0 не отвергается => параметры модели незначимы



In [ ]:
F_stat = 68.74 # F_набл
F_fr = qf(0.975, 2, 72) # F_табл(0.05, 2, 72)
paste0(round(F_fr,3))

[1] "3.885"

> F_набл > F_табл => гипотеза H0 отвергается => модель значима

**t-критерий Стьюдента для проверки значимости коэффициента корреляции:**

In [ ]:
t_fr = qt(0.975, 73) # t_табл(0.05, 73) 
paste0(round(t_fr,3))

[1] "1.993"

r_x1_x2

In [ ]:
result = cor.test(df[,2], df[,3], method = "pearson")
print(result)


	Pearson's product-moment correlation

data:  df[, 2] and df[, 3]
t = 7.2732, df = 73, p-value = 3.235e-10
alternative hypothesis: true correlation is not equal to 0
95 percent confidence interval:
 0.4939078 0.7629280
sample estimates:
      cor 
0.6482068 



### **Вывод: значим**


r_x1_y

In [ ]:
result = cor.test(df[,2], df[,1], method = "pearson")
print(result)


	Pearson's product-moment correlation

data:  df[, 2] and df[, 1]
t = 11.733, df = 73, p-value < 2.2e-16
alternative hypothesis: true correlation is not equal to 0
95 percent confidence interval:
 0.7120688 0.8748384
sample estimates:
      cor 
0.8083855 



### **Вывод: значим**

r_x2_y

In [ ]:
result = cor.test(df[,3], df[,1], method = "pearson")
print(result)


	Pearson's product-moment correlation

data:  df[, 3] and df[, 1]
t = 5.8404, df = 73, p-value = 1.343e-07
alternative hypothesis: true correlation is not equal to 0
95 percent confidence interval:
 0.3869150 0.7014426
sample estimates:
     cor 
0.564321 



### **Вывод: значим**

#**Коэффициенты бетта, дельта и эластичности.**

In [ ]:
summary(model_1)


Call:
lm(formula = y ~ x1 + x2, data = df)

Residuals:
     Min       1Q   Median       3Q      Max 
-1.28332 -0.31502 -0.08199  0.38497  1.72936 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) 16.55736   29.18175   0.567    0.572    
x1           0.73504    0.08737   8.412  2.6e-12 ***
x2           0.07590    0.09904   0.766    0.446    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.6118 on 72 degrees of freedom
Multiple R-squared:  0.6563,	Adjusted R-squared:  0.6467 
F-statistic: 68.74 on 2 and 72 DF,  p-value: < 2.2e-16


## Средний коэффициент эластичности

In [ ]:
E_x1 = 0.73504 * (mean(df[,2])/mean(df[,1])) # Средний коэффициент эластичности для x1
round(E_x1,3)

[1] 0.931

### С увеличением x1 на 1%, Y увеличится на 0.931%

In [ ]:
E_x2 = 0.07590 * (mean(df[,3])/mean(df[,1])) # Средний коэффициент эластичности для x2
round(E_x2,3)

[1] 0.01

### С увеличением x2 на 1%, Y увеличится на 0.01%

Delta коэф для x1

In [ ]:
cor_x1_y = cor(df[,1], df[,2])
cor_x2_y = cor(df[,1], df[,3])

In [ ]:
delta_coef_x1 = cor_x1_y*(0.73504/0.6563)
round(delta_coef_x1,3)

[1] 0.905

Delta коэф для x2

In [ ]:
delta_coef_x2 = cor_x2_y*(0.07590/0.6563)
round(delta_coef_x2,3)

[1] 0.065

### Дельта коэф показывает среднюю долю влияния соответствующего фактора в совокупном влиянии всех факторов, включённых в модель.
### Таким образом x1 оказывает наибольшую долю влияния во множественной моделе регрессии 

##**Итоговый вывод по модели: Так как коэфф детерминации и скорректировннаый коэфф детерминации равны 0.656 и 0.647 соответсвенно и оба меньше 0.7, а средняя относительная ошибка аппроксимации (0.163 %) меньше 5% , то качество модели среднее.**

#**Сопоставьте длинную и короткую модели, в качестве короткой модели используйте линейную модель с фактором, который имеют более сильную связь с эндогенной переменной:**

In [ ]:
cor_x1_y = cor(df[,1], df[,2])
cor_x1_y

[1] 0.8083855

In [ ]:
cor_x2_y = cor(df[,1], df[,3])
cor_x2_y

[1] 0.564321

Из полученных парных коэффициентов корреляции можно сделать вывод, что наиболее коррелирует y с фактором x1 => будем строить короткую модель с x1

In [ ]:
model_2 = lm(y ~ x1, data=df)

In [ ]:
summary(model_2)


Call:
lm(formula = y ~ x1, data = df)

Residuals:
     Min       1Q   Median       3Q      Max 
-1.28989 -0.33213 -0.06038  0.43339  1.69660 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  3.85291   23.94801   0.161    0.873    
x1           0.77844    0.06634  11.733   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.6101 on 73 degrees of freedom
Multiple R-squared:  0.6535,	Adjusted R-squared:  0.6487 
F-statistic: 137.7 on 1 and 73 DF,  p-value: < 2.2e-16


### Тест на длинную/короткую модель регрессии:

In [ ]:
waldtest(model_2, model_1, test = c("F"))

Res.Df Df F         Pr(>F)   
1 73     NA        NA        NA
2 72      1 0.5873335 0.4459576

### Т.к. P-value > alfa (0.4459576 > 0.05) => H0 верна => Предпочтительнее короткая модель регрессии с x1 


**Выбранную модель проверяем на наличие пропущенных переменных.**

**Используя тест Рамсея, проверим нашу модель**

Строим вспомогательную модель регрессиии от x1 

$$ \hat{\hat{y_i}} =a+b_1+x_{1i}+b_2\hat{y_i}^2 +b_3\hat{y_i}^3$$

$$\hat{y_i} =a+b_1+x_{1i}$$

---

$$ H_0: b_2=b_3=0 $$

$$ H_1: b_2^2=b_3^2=0 $$


In [ ]:
resettest(model_2, type="regressor")


	RESET test

data:  model_2
RESET = 0.0037649, df1 = 2, df2 = 71, p-value = 0.9962


$$ \text{Т.к. } p_{value} > alfa \text{     (0.9962 > 0.05)} => {H_0} \text{ - верна} => \text{в модели нет пропущенных факторов} $$


In [ ]:
predict(model_2, newdata = df, interval = "confidence") # точечный

fit      lwr      upr     
1  285.5342 285.3509 285.7174
2  284.7196 284.5777 284.8615
3  285.3199 285.1576 285.4821
4  285.7123 285.5083 285.9163
5  283.8855 283.6709 284.1001
6  285.3574 285.1919 285.5230
7  284.6635 284.5199 284.8071
8  285.0977 284.9507 285.2447
9  285.5767 285.3887 285.7646
10 284.8966 284.7559 285.0373
11 284.1013 283.9129 284.2897
12 284.0559 283.8622 284.2495
13 284.1297 283.9445 284.3149
14 284.8619 284.7214 285.0023
15 284.2980 284.1300 284.4660
16 285.5163 285.3350 285.6976
17 284.9131 284.7722 285.0541
18 284.5119 284.3608 284.6630
19 284.6051 284.4590 284.7511
20 285.3868 285.2185 285.5550
21 284.2365 284.0625 284.4104
22 285.7177 285.5130 285.9224
23 285.1848 285.0327 285.3368
24 284.1220 283.9359 284.3081
25 285.3679 285.2014 285.5344
26 284.6325 284.4877 284.7773
27 284.4891 284.3365 284.6417
28 282.9833 282.6380 283.3287
29 286.2447 285.9680 286.5214
30 284.6793 284.5363 284.8224
⋮  ⋮        ⋮        ⋮       
46 284.9611 284.8192 285.1030
47 284.3003 284.1325 284.4681
48 283.4057 283.1243 283.6870
49 284.6223 284.4770 284.7675
50 284.6435 284.4992 284.7879
51 283.4061 283.1248 283.6874
52 285.1227 284.9743 285.2710
53 285.6419 285.4464 285.8374
54 285.2338 285.0784 285.3893
55 283.2630 282.9604 283.5655
56 284.5344 284.3846 284.6841
57 285.9783 285.7395 286.2171
58 284.6882 284.5455 284.8310
59 286.1799 285.9126 286.4471
60 284.7818 284.6411 284.9226
61 284.2504 284.0778 284.4229
62 284.4666 284.3124 284.6207
63 285.6125 285.4205 285.8046
64 284.3603 284.1979 284.5227
65 285.8946 285.6671 286.1220
66 284.3801 284.2194 284.5409
67 284.0351 283.8390 284.2311
68 285.5197 285.3380 285.7013
69 286.0385 285.7914 286.2857
70 285.8205 285.6028 286.0382
71 285.1104 284.9627 285.2581
72 284.3203 284.1544 284.4863
73 284.5668 284.4189 284.7147
74 284.1279 283.9424 284.3133
75 286.5303 286.2109 286.8498